src/python/pants/backend/python/subsystems/pex_bootstrap/_ptex_launcher.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# This bootstrapping technique for lazily loading requirements comes from the pex project test case:
+# https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+
+import json
+import os
+import sys
+
+from pex import resolver
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+from pex.platforms import Platform
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+hydrated_ipex_file = '{}.ipex'.format(filename_base)
+
+if not os.path.isfile(hydrated_ipex_file):
+  print('Hydrating {} to {}'.format(self, hydrated_ipex_file))
+
+  wrapper_ptex_info = PexInfo.from_pex(self)
+  if wrapper_ptex_info.zip_safe:
+    raise ValueError("Could not hydrate ptex: ipex info must be marked zip_safe=False. This ipex file's info was:\\n{}".format(wrapper_ptex_info.dump(indent=4)))
+  code_root = os.path.join(wrapper_ptex_info.zip_unsafe_cache, wrapper_ptex_info.code_hash)
+  with open_zip(self) as zf:
+    # Populate the pex with the pinned requirements and distribution names & hashes.
+    ipex_info = PexInfo.from_json(zf.read('IPEX-INFO'))
+    ipex_builder = PEXBuilder(pex_info=ipex_info)
+
+    # Populate the pex with the needed code.
+    ptex_info = json.loads(zf.read('PTEX-INFO').decode('utf-8'))
+    for path in ptex_info['code']:
+      ipex_builder.add_source(os.path.join(code_root, path), path)
+
+    # Get the requirements to resolve.
+    plat = Platform.current()
+    transitive_requirements_by_platform = ptex_info['transitive_requirements_by_platform']
+    if plat.platform not in transitive_requirements_by_platform:
+      raise ValueError("Current platform {} was not found in the 'transitive_requirements' field in PTEX-INFO:\\n{}".format(plat, json.dumps(ptex_info, indent=4)))
+
+  # Perform a fully pinned intransitive resolve to hydrate the install cache (not the
+  # pex!).
+  resolver_settings = ptex_info['resolver_settings']
+  # TODO: Convert .indexes and .find_links into the old .fetchers until pants upgrades to pex 2.0!
+  fetchers = [PyPIFetcher(url) for url in resolver_settings['indexes']]
+  del resolver_settings['indexes']
+  fetchers.extend(Fetcher([url]) for url in resolver_settings['find_links'])
+  del resolver_settings['find_links']
+  resolver_settings['fetchers'] = fetchers
+
+  resolved_distributions = resolver.resolve(
+    requirements=transitive_requirements_by_platform[plat.platform],
```

Comment:
I'll re-iterate that this is not required in case it got lost in the shuffle (please accept my apologies if not). You can & should use the ipex_info.requirements verbatim.

Reply:
This was super helpful! Thank you!